### PR TITLE
interop-testing: Add concurrency condition to the soak test using existing blocking api 

### PR DESCRIPTION
**Pull request by zbilun**
The goal of this PR is to increase the test coverage of the C2P E2E load test by improving the rpc_soak and channel_soak tests to support concurrency.

**rpc_soak:**
The client performs many large_unary RPCs in sequence over the same channel. The test can run in either a concurrent or non-concurrent mode, depending on the number of threads specified (soak_num_threads):
  - Non-Concurrent Mode: When soak_num_threads = 1, all RPCs are performed sequentially on a single thread.
  - Concurrent Mode: When soak_num_threads > 1, the client uses multiple threads to distribute the workload. Each thread performs a portion of the total soak_iterations, executing its own set of RPCs concurrently.

**channel_soak:**
Similar to rpc_soak, but this time each RPC is performed on a new channel. The channel is created just before each RPC and is destroyed just after. Note on Concurrent Execution and Channel Creation: In a concurrent execution setting (i.e., when soak_num_threads > 1), each thread performs a portion of the total soak_iterations and creates and destroys its own channel for each RPC iteration.
- createNewChannel Function: In channel_soak, the createNewChannel function is used by each thread to create a new channel before every RPC. This function ensures that each RPC has a separate channel, preventing race conditions by isolating channels between threads. It shuts down the previous channel (if any) and creates a new one for each iteration, ensuring accurate latency measurement per RPC.

- Thread-specific logs will include the thread_id, helping to track performance across threads, especially when each thread is managing its own channel lifecycle.

PTAL @apolcyn 
@DNVindhya

### CODE REVIEW

File: interop-testing/src/main/java/io/grpc/testing/integration/AbstractInteropTest.java
Comment (apolcyn) [inline interop-testing/src/main/java/io/grpc/testing/integration/AbstractInteropTest.java:1757]: Rather than hardcoding 10 let's make parallelism configurable with a flag.

We can add a new flag, something like `--soak_num_threads`, where the value of the flag determines the number of number of threads to independently run the soak test loop on.

The flag can default to `1`.

File: interop-testing/src/main/java/io/grpc/testing/integration/TestServiceClient.java
Comment (apolcyn) [inline interop-testing/src/main/java/io/grpc/testing/integration/TestServiceClient.java:534]: Rather than introduce new test cases, let's reuse the existing `rpc_soak` and `channel_soak` test cases, but add one more parameter for number of threads (passing the value in from the aforementioned `--soak_num_threads` flag)

File: interop-testing/src/main/java/io/grpc/testing/integration/AbstractInteropTest.java
Comment (apolcyn) [inline interop-testing/src/main/java/io/grpc/testing/integration/AbstractInteropTest.java:1775]: Rather than create a new thread per iteration, I think it would be better to create `--num_threads` threads at the beginning of the soak test, and then have each thread independently run the soak test loop (e.g. run `--soak_iterations` RPCs).

File: interop-testing/src/main/java/io/grpc/testing/integration/AbstractInteropTest.java
Comment (zbilun) [reply interop-testing/src/main/java/io/grpc/testing/integration/AbstractInteropTest.java]: - I have updated the code based on your comments. Currently, I am implementing a uniform distribution for the soak iterations across different threads. However, there are several other approaches to distribute the soak iterations dynamically, such as round robin or load balancing. Please let me know if you would prefer a different implementation.
- Also I tried to run "tools/distrib/sanitize.sh" to format the code, however i could not find any sanitize.sh file from the root directory. Any hint?
- I may also need to modify the related .py files. I will update those files and submit a change list (CL) for review afterward.

File: interop-testing/src/main/java/io/grpc/testing/integration/AbstractInteropTest.java
Comment (apolcyn) [inline interop-testing/src/main/java/io/grpc/testing/integration/AbstractInteropTest.java:1753]: Let's check these flags and fail at startup (before running the test), if --soak_iterations is not evenly divisible by --num_threads.

Then, let's just have each thread independently run `soakIterationsPerThread` RPCs (no need to calculate `startIteration` and `endIteration`)

File: interop-testing/src/main/java/io/grpc/testing/integration/AbstractInteropTest.java
Comment (apolcyn) [inline interop-testing/src/main/java/io/grpc/testing/integration/AbstractInteropTest.java:1783]: we shouldn't need to examine `totalFailures` until the thread is done and joined.

We can have a list of per-thread counters that each thread tallies up while running. Then after joining all threads, the main thread can examine all of those counters and create a summary.

File: interop-testing/src/main/java/io/grpc/testing/integration/AbstractInteropTest.java
Comment (apolcyn) [inline interop-testing/src/main/java/io/grpc/testing/integration/AbstractInteropTest.java:1758]: nit: let's put the thread body in its own function

File: interop-testing/src/main/java/io/grpc/testing/integration/AbstractInteropTest.java
Comment (apolcyn) [inline interop-testing/src/main/java/io/grpc/testing/integration/AbstractInteropTest.java:1773]: nit: the way that we're resetting the channel here might be problematic here, because, for example, on the first iteration, all threads will shut down the same channel.

I think we need to have all threads start out initially with an independent channel.

File: interop-testing/src/main/java/io/grpc/testing/integration/AbstractInteropTest.java
Comment (apolcyn) [inline interop-testing/src/main/java/io/grpc/testing/integration/AbstractInteropTest.java:1732]: Rather than take a `resetChannelPerIteration` parameter, I think we can clean things up by taking a function parameter here instead.

The signature of the function param can be like this: `ManagedChannel maybeCreateNewChannel(ManagedChannel channel)`

In the rpc soak test, we can provide a callback that always returns the same channel.

In the channel soak test, we can provide a callback that closes the pass-in channel and returns a new one.

File: interop-testing/src/main/java/io/grpc/testing/integration/AbstractInteropTest.java
Comment (apolcyn) [inline interop-testing/src/main/java/io/grpc/testing/integration/AbstractInteropTest.java:1766]: There are several pieces of data that are *output* from each of these threads:

- `threadFailures`
- `iterationsDone`
- `latencies`

Rather than pass each of these as separate atomic parameters, I think it would be cleaner to pass a single parameter to each of the threads. We can create a simple wrapper object to wrap each of `threadFailures`, `iterationsDone`, and `latencies`.

In the main thread here, we can create a list of such wrapper objects.

That way, each thread has a separate copy of mutable output data, and there is no need to synchronize anything.

After all threads are joined, the main thread can merge results across the list.

**Review by apolcyn — APPROVED** [review]
LGTM after comments are addressed

(I'll be OOO tomorrow and monday)

File: interop-testing/src/main/java/io/grpc/testing/integration/AbstractInteropTest.java
Comment (zbilun) [reply interop-testing/src/main/java/io/grpc/testing/integration/AbstractInteropTest.java:1795]: Updated the function based on the comment. However, as lambda expressions and the Thread class don't allow checked exceptions (like InterruptedException) to be propagated directly, I add the try catch here " threads[threadInd] = new Thread(() -> {
        try {
          executeSoakTestInThreads("

**Comment by zbilun** [comment]
In the TestServiceClient.java,  case RPC_SOAK and case CHANNEL_SOAK, I have added the try catch here due to the IDE's error about the unhandled InterruptedException in the lambda.

**Comment by zbilun** [comment]
During our last meeting, you mentioned that we don't need to modify the code in XdsFederationTestClient.java. However, in the PR, we've made changes to the performSoakTest which is called in the XdsFederationTestClient  I think we may need to update the calling code in XdsFederationTestClient.java. Any thoughts on this?

File: interop-testing/src/main/java/io/grpc/testing/integration/AbstractInteropTest.java
Comment (zbilun) [reply interop-testing/src/main/java/io/grpc/testing/integration/AbstractInteropTest.java]: A separate pull request to update the interop test documentation will be submitted once this PR is approved.

File: interop-testing/src/main/java/io/grpc/testing/integration/AbstractInteropTest.java
Comment (apolcyn) [inline interop-testing/src/main/java/io/grpc/testing/integration/AbstractInteropTest.java:1779]: nit: `s/executeSoakTestInThreads/executeSoakTestInThread`

File: interop-testing/src/main/java/io/grpc/testing/integration/AbstractInteropTest.java
Comment (apolcyn) [inline interop-testing/src/main/java/io/grpc/testing/integration/AbstractInteropTest.java:1792]: can we re-throw this as a `RuntimeException` to get around the checked exception / thread function restriction?

**Comment by apolcyn** [comment]
>  During our last meeting, you mentioned that we don't need to modify the code in XdsFederationTestClient.java. However, in the PR, we've made changes to the performSoakTest which is called in the XdsFederationTestClient I think we may need to update the calling code in XdsFederationTestClient.java. Any thoughts on this?


Oh in that case we should just apply the same change to `XdsFederationTestClient` as we did for `TestServiceClient`: instead of passing a boolean to reset the channel or not, pass the appropriate lambda

File: interop-testing/src/main/java/io/grpc/testing/integration/AbstractInteropTest.java
Comment (zbilun) [reply interop-testing/src/main/java/io/grpc/testing/integration/AbstractInteropTest.java:1792]: Unfortunately, I have tried re-throwing the InterruptedException as a RuntimeException, but it doesn't resolve the issue.

Comment (zbilun) [reply interop-testing/src/main/java/io/grpc/testing/integration/AbstractInteropTest.java:1791]: I think I misunderstood your point earlier. You are correct — we can re-throw the InterruptedException as a RuntimeException, and I've made that change.

Based on my study today, regarding lambda exception handling: The try-catch block inside the lambda is required because a lambda cannot directly throw checked exceptions. If the lambda encounters an InterruptedException, we catch it and re-throw it as an unchecked RuntimeException. Please let me know if this matches what you meant!

File: interop-testing/src/main/java/io/grpc/testing/integration/TestServiceClient.java
Comment (apolcyn) [inline interop-testing/src/main/java/io/grpc/testing/integration/TestServiceClient.java:560]: nit: we might just have `maybeCreateNewChannel` itself catch `InterruptedException` and re-throw as `RuntimeException` (so that . That way `maybeCreateNewChannel` no longer throws a checked exception, so we don't need to re-type this try/catch in every lambda we pass to `performSoakTest`

File: interop-testing/src/main/java/io/grpc/testing/integration/XdsFederationTestClient.java
Comment (apolcyn) [inline interop-testing/src/main/java/io/grpc/testing/integration/XdsFederationTestClient.java:79]: nit: we don't need to make numThreads configurable for `XdsFederationTestClient`, we can leave it hardcoded at 1.

I.e. we can remove the new flag here

File: interop-testing/src/main/java/io/grpc/testing/integration/AbstractInteropTest.java
Comment (apolcyn) [inline interop-testing/src/main/java/io/grpc/testing/integration/AbstractInteropTest.java:1802]: no need to use `AtomicInteger` here (we're not accessing `iterationsDone` concurrently)

let's just use `int`

File: interop-testing/src/main/java/io/grpc/testing/integration/AbstractInteropTest.java
Comment (apolcyn) [inline interop-testing/src/main/java/io/grpc/testing/integration/AbstractInteropTest.java:115]: nit: we can remove this import not

Same with `AtomicInteger`

File: interop-testing/src/main/java/io/grpc/testing/integration/AbstractInteropTest.java
Comment (apolcyn) [inline interop-testing/src/main/java/io/grpc/testing/integration/AbstractInteropTest.java:690]: we shouldn't be changing this?

File: interop-testing/src/main/java/io/grpc/testing/integration/XdsFederationTestClient.java
Comment (apolcyn) [inline interop-testing/src/main/java/io/grpc/testing/integration/XdsFederationTestClient.java:263]: nit: we can remove this try/catch now, same below

File: interop-testing/src/main/java/io/grpc/testing/integration/TestServiceClient.java
Comment (apolcyn) [inline interop-testing/src/main/java/io/grpc/testing/integration/TestServiceClient.java:536]: nit: we might actually make this more readable if we removed the boolean parameter from `tester.maybeCreateNewChannel`

I.e.:

1) rename `tester.maybeCreateNewChannel` to `tester.createNewChannel`

2) In rpc_soak cases, just pass a lambda like this: `(currentChannel) -> currentChannel`

3) In channel_soak cases, pass `(currentChannel) -> tester.createNewChannel(currentChannel)`

**Review by apolcyn — APPROVED** [review]

File: interop-testing/src/main/java/io/grpc/testing/integration/XdsFederationTestClient.java
Comment (apolcyn) [inline interop-testing/src/main/java/io/grpc/testing/integration/XdsFederationTestClient.java:130]: this line accidentally changed?

**Review by apolcyn — APPROVED** [review]

File: interop-testing/src/main/java/io/grpc/testing/integration/XdsFederationTestClient.java
Comment (zbilun) [reply interop-testing/src/main/java/io/grpc/testing/integration/XdsFederationTestClient.java:130]: yes....just fixed it.

**Review by DNVindhya — APPROVED** [review]

**Pull request merged**
